src/composer/ComposerOracle.sol

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity >=0.8.0;
+
+import {IOracle} from "../interfaces/IOracle.sol";
+import {BoringERC20} from "../libraries/BoringERC20.sol";
+import {IERC20} from "forge-std/interfaces/IERC20.sol";
+
+import { console2 } from "forge-std/console2.sol";
+
+/**
+ * @title ComposerOracle
+ */
+contract ComposerOracle is IOracle {
+
+    event OracleSet(address indexed base, address indexed quote, IOracle indexed oracle);
+    event PathSet(address indexed base, address indexed quote, address[] indexed path);
+
+    struct OracleWithDecimals {
+        IOracle oracle;
+        uint8 baseDecimals;
+        uint8 quoteDecimals;
+    }
+
+    struct SetOracle {
+        address base;
+        address quote;
+        OracleWithDecimals oracle;
+    }
+
+    struct SetPath {
+        address base;
+        address quote;
+        address[] path;
+    }
+
+    mapping(address => mapping(address => OracleWithDecimals)) public oracles;
+    mapping(address => mapping(address => address[])) public paths;
+
+    constructor (SetOracle[] memory oracles_, SetPath[] memory paths_) {
+        uint256 oraclesLength = oracles_.length;
+        unchecked {
+            for (uint256 p; p < oraclesLength; ++p) {
+                _setOracle(oracles_[p].base, oracles_[p].quote, oracles_[p].oracle);
+            }
+        }
```

Comment:
now would be a good time to decide a common solidity version across the repo and pin it via `foundry.toml`.

this `unchecked` block is perfect with `>=0.8.0` it is redundant in versions after `0.8.22` due to [compiler optimisations](https://soliditylang.org/blog/2023/10/25/solidity-0.8.22-release-announcement/).

i suggest we bump the solc version to 0.8.24 across the repo (latest is 0.8.25) for now.

Reply:
That's a good point. I'd leave >=0.8.0 as the pragma for all contracts, and decide elsewhere which version to use. It might be chain-dependent.

So I guess that the `unchecked` block stays because there are chains that don't support solc >=0.8.22

Reply:
good idea, i'll keep this in mind for future oracles too